src/azul/tdr.py

```diff
@@ -0,0 +1,52 @@
+from collections import Counter
+from functools import cached_property
+from typing import (
+    List,
+    NamedTuple,
+)
+
+from azul.bigquery import (
+    AbstractBigQueryAdapter,
+    BigQueryAdapter,
+)
+from azul.indexer import BundleFQID
+from azul.uuids import validate_uuid_prefix
+
+
+class BigQueryDataset(NamedTuple):
+    id: str
+    name: str
+    is_snapshot: bool
+
+    @property
+    def bigquery_dataset_name(self):
+        return self.name if self.is_snapshot else f'datarepo_{self.name}'
+
+
+class AzulTDRClient:
+    timestamp_format = '%Y-%m-%dT%H:%M:%S.%fZ'
+
+    @cached_property
+    def big_query_adapter(self) -> AbstractBigQueryAdapter:
+        return BigQueryAdapter()
+
+    def __init__(self, dataset: BigQueryDataset):
+        self.target = dataset
+        self.big_query_adapter.assert_table_exists(dataset.bigquery_dataset_name, 'links')
+
+    def list_links_ids(self, prefix: str) -> List[BundleFQID]:
+        validate_uuid_prefix(prefix)
+        iter_rows = self.big_query_adapter.run_sql(f'''
+            SELECT links_id, version
+            FROM {self.target.bigquery_dataset_name}.links
+            WHERE STARTS_WITH(links_id, "{prefix}")
+        ''')
+        bundles_fqids = [BundleFQID(uuid=row['links_id'],
+                                    version=row['version'].strftime(self.timestamp_format))
+                         for row in iter_rows]
+
+        if self.target.is_snapshot:
```

Comment:
Looking at the `azulclient` module I now realize that only the most recent version should be included in the list of bundles when multiple versions are present. I'm refactoring this check to a separate method since it's the same pattern that's used extensively in the `fetch_bundle` implementation.

Reply:
Actually, this is redundant since `AzulClient` still filters out non-current bundles, but my refactor has the added advantage of verifying bundle UUID uniqueness, which Terra only enforces during ingest due to lack of proper keys in BigQuery. Might need to revert though if there's a reason why `list_bundles` should include outdated versions.

Reply:
Added a unit test to cover the separate snapshot and dataset cases